src/lib/promise/js/promise.ml

```diff
@@ -0,0 +1,38 @@
+type 'a t
+
+external run_in_thread : (unit -> 'a) -> 'a t = "deferred_run"
+
+let block_on_async_exn (_ : unit -> 'a t) : 'a =
+  failwith "You can't block on async execution in JS"
+
+external map : 'a t -> f:('a -> 'b) -> 'b t = "deferred_map"
+
+external bind : 'a t -> f:('a -> 'b t) -> 'b t = "deferred_bind"
+
+external upon : 'a t -> ('a -> unit) -> unit = "deferred_upon"
+
+external is_determined : 'a t -> bool = "deferred_is_determined"
+
+external peek : 'a t -> 'a option = "deferred_peek"
+
+external value_exn : 'a t -> 'a = "deferred_value_exn"
+
+external return : 'a -> 'a t = "deferred_return"
+
+external create : (('a -> unit) -> unit) -> 'a t = "deferred_create"
+
+let to_deferred promise =
+  let module Ivar = Async_kernel.Ivar in
+  let ivar = Ivar.create () in
+  upon promise (fun x -> Ivar.fill ivar x) ;
```

Comment:
I know I suggested this, but I don't think it'll handle exceptions properly, since `upon` will have swallowed the exception. We probably need an `upon_exn`, and to call this `to_deferred_exn`.

Reply:
@mrmr1993 As currently implemented, when the promise rejects, this will throw an error, which cannot be caught in any way (because upon creates a new promise, which will reject when it's parent  promise rejects, but the new promise doesn't get exposed so it can't be caught)

Reply:
So this is not ideal obiously, but I'm wondering: How are exceptions in deferreds usually handled? I'd want to emulate that

Reply:
They get caught by the scheduler, so the point at which we attach the promise to the scheduler feels like the right time to raise. The current implementation is really `upon_exn` (which I missed in review), and the real `upon` can be
```js
function deferred_upon(deferred, func) {
  deferred.promise.then(function () {
    func(deferred.value);
  }).catch(function(){});
}
```

Reply:
I changed the upon_exn / upon thing but I'm hesitant to call this function to_deferred_exn, since to_deferred can't throw an exception in native ocaml and this JS version was actually just supposed to be a dummy function (we don't have a way of using Deferred in JS after all). So I'd say this function, with `upon` called `upon_exn`, is "good enough" for now